### PR TITLE
executor: reorg codes for hashtable in HashJoinExec

### PR DESCRIPTION
**Pull request by SunRunAway**
### What problem does this PR solve? 

Part of #11607
It reorgs codes for the hash table and introduces `hashRowContainer`. So that In further pull request, I can implement `spillOutToDisk` feature for `hashRowContainer`

### What is changed and how it works?

- [x] reorg codes for hash table in `HashJoinExec`
- [x] make a suitable `maxEntrySliceLen`, it's currently too large(8k) for OLTP queries.
- [x] add unit test for `rowHashMap`.
- [x] try to replace FNV64 to another hash function. (abandon, FNV64 is the fastest.)
- [x] initialize the size of `rowHashMap` from the estimated row count for better performance.

### Check List 

Tests 

 - Unit test

Code changes

 - None

Side effects

 - None

Related changes

 - None

Release note

 - None

### Benchmark

```
name                                                                     old time/op    new time/op    delta
HashJoinExec/(rows:100000,_concurency:4,_joinKeyIdx:_[0_1])-12              760ms ± 6%     781ms ± 7%     ~     (p=0.421 n=5+5)
HashJoinExec/(rows:100000,_concurency:4,_joinKeyIdx:_[0])-12                168ms ± 5%     166ms ± 2%     ~     (p=0.841 n=5+5)
BuildHashTableForList/(rows:100000,_concurency:4,_joinKeyIdx:_[0_1])-12     512ms ± 0%     544ms ± 1%   +6.19%  (p=0.016 n=4+5)
BuildHashTableForList/(rows:100000,_concurency:4,_joinKeyIdx:_[0])-12      12.2ms ± 2%    12.5ms ± 8%     ~     (p=0.548 n=5+5)
BuildHashTableForList/(rows:10,_concurency:4,_joinKeyIdx:_[0])-12          2.46µs ± 2%    2.94µs ± 0%  +19.46%  (p=0.008 n=5+5)

name                                                                     old alloc/op   new alloc/op   delta
HashJoinExec/(rows:100000,_concurency:4,_joinKeyIdx:_[0_1])-12              305MB ± 0%     304MB ± 0%   -0.28%  (p=0.008 n=5+5)
HashJoinExec/(rows:100000,_concurency:4,_joinKeyIdx:_[0])-12                305MB ± 0%     304MB ± 0%   -0.28%  (p=0.008 n=5+5)
BuildHashTableForList/(rows:100000,_concurency:4,_joinKeyIdx:_[0_1])-12    7.97MB ± 0%    7.11MB ± 0%  -10.80%  (p=0.008 n=5+5)
BuildHashTableForList/(rows:100000,_concurency:4,_joinKeyIdx:_[0])-12      7.97MB ± 0%    7.11MB ± 0%  -10.80%  (p=0.008 n=5+5)
BuildHashTableForList/(rows:10,_concurency:4,_joinKeyIdx:_[0])-12          1.62kB ± 0%    1.96kB ± 0%  +20.74%  (p=0.008 n=5+5)

name                                                                     old allocs/op  new allocs/op  delta
HashJoinExec/(rows:100000,_concurency:4,_joinKeyIdx:_[0_1])-12               306k ± 0%      306k ± 0%   -0.08%  (p=0.008 n=5+5)
HashJoinExec/(rows:100000,_concurency:4,_joinKeyIdx:_[0])-12                 306k ± 0%      306k ± 0%   -0.09%  (p=0.008 n=5+5)
BuildHashTableForList/(rows:100000,_concurency:4,_joinKeyIdx:_[0_1])-12     4.07k ± 2%     3.80k ± 1%   -6.56%  (p=0.008 n=5+5)
BuildHashTableForList/(rows:100000,_concurency:4,_joinKeyIdx:_[0])-12       4.04k ± 0%     3.79k ± 0%   -6.41%  (p=0.008 n=5+5)
BuildHashTableForList/(rows:10,_concurency:4,_joinKeyIdx:_[0])-12            9.00 ± 0%     14.00 ± 0%  +55.56%  (p=0.008 n=5+5)
```
[old.txt](https://github.com/pingcap/tidb/files/3559836/old.txt)
[new.txt](https://github.com/pingcap/tidb/files/3559840/new.txt)

### CODE REVIEW

File: executor/hash_table.go [hunk 1]
```diff
@@ -56,23 +200,25 @@ var nullEntryAddr = entryAddr{}
...
+// TODO(fengliyuan): add unit test for this.
 type rowHashMap struct {
 	entryStore entryStore
 	hashTable  map[uint64]entryAddr
 	length     int
 }
 
 // newRowHashMap creates a new rowHashMap.
-func newRowHashMap() *rowHashMap {
+func newRowHashMapWithStatCount(statCount int) *rowHashMap {
```
Comment (zz-jason) [inline executor/hash_table.go]: Could you explain what does `statCount` mean? BTW, the comment for this function should be started with `newRowHashMapWithStatCount`.

Comment (qw4990) [inline executor/hash_table.go]: `statCount` -> `initCap`?

File: executor/hash_table.go [hunk 2]
```diff
@@ -14,31 +14,175 @@
...
+// TODO: support spilling out to disk when memory is limited.
+type hashRowContainer struct {
+	records   *chunk.List
+	hashTable *rowHashMap
+
+	sc        *stmtctx.StatementContext
+	allTypes  []*types.FieldType
+	keyColIdx []int
+	h         hash.Hash64
+	buf       [1]byte
```
[the remark this answers is not in the document]
Comment (SunRunAway) [reply executor/hash_table.go]: `[1]byte` is more clearly demonstrating that it has only 1 byte and need not slice initialization.

File: executor/hash_table.go [hunk 3]
```diff
@@ -14,31 +14,175 @@
...
+		err     error
+		key     uint64
+	)
+	numRows := chk.NumRows()
+	for j := 0; j < numRows; j++ {
+		hasNull, key, err = c.getJoinKeyFromChkRow(c.sc, chk.GetRow(j), c.allTypes, c.keyColIdx, c.h, c.buf[:])
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if hasNull {
```
Comment (XuHuaiyu) [inline executor/hash_table.go]: Should we record the `rowIdx` of which `hasNull` is true?
If we use the outer table to build hashtable, we need to know the null-value lines.

File: executor/hash_table.go [hunk 4]
```diff
@@ -14,31 +14,175 @@
...
+
+func (c *hashRowContainer) GetMemTracker() *memory.Tracker {
+	return c.records.GetMemTracker()
+}
+
+// GetMatchedRows get matched rows from probeRow. It can be called
+// in multiple goroutines while each goroutine should keep its own
+// h and buf.
+func (c *hashRowContainer) GetMatchedRows(probeRow chunk.Row, joinKeysTypes []*types.FieldType, keyColIdx []int, h hash.Hash64, buf []byte) (matched []chunk.Row, hasNull bool, err error) {
+
```
Comment (XuHuaiyu) [inline executor/hash_table.go]: useless line

File: executor/hash_table.go [hunk 5]
```diff
@@ -14,31 +14,175 @@
...
+	hasNull, key, err = c.getJoinKeyFromChkRow(c.sc, probeRow, joinKeysTypes, keyColIdx, h, buf)
+	if err != nil {
+		return
+	}
+	if hasNull {
+		return
+	}
+	innerPtrs := c.hashTable.Get(key)
+	if len(innerPtrs) == 0 {
+		hasNull = true
```
Comment (XuHuaiyu) [inline executor/hash_table.go]: Why we set `hashNull` to `true`?

File: executor/hash_table.go [hunk 6]
```diff
@@ -14,31 +14,175 @@
...
+		if err != nil {
+			return
+		}
+		if !ok {
+			continue
+		}
+		matched = append(matched, matchedRow)
+	}
+	if len(matched) == 0 { // TODO(fengliyuan): add test case
+		hasNull = true
```
Comment (XuHuaiyu) [inline executor/hash_table.go]: ditto

File: executor/hash_table.go (hunk 5 above)
Comment (SunRunAway) [reply executor/hash_table.go]: It means no inner rows matched by the outer row in `hashRowContainer` so we need call `onMissMatch`.
The same logic in old code is at https://github.com/pingcap/tidb/blob/master/executor/join.go#L417 and https://github.com/pingcap/tidb/blob/master/executor/join.go#L434

File: executor/hash_table.go (hunk 3 above)
Comment (SunRunAway) [reply executor/hash_table.go]: I'll adjust it in further PR to involve hashRowContainer into index join.

**Comment by SunRunAway** [comment]
All comments are addressed.
PTAL @zz-jason, @qw4990, @XuHuaiyu

File: executor/hash_table.go (hunk 5 above)
Comment (SunRunAway) [reply executor/hash_table.go]: I've addressed this by removing the ambiguous `isNull` value, and use the size of matched rows to call `onMissMatch`.

File: executor/hash_table.go [hunk 7]
```diff
@@ -14,31 +14,199 @@
...
+	if len(matched) == 0 {
+		// noop
+	}
+	*/
+	return
+}
+
+// matchJoinKey checks if join keys of buildRow and probeRow are logically equal.
+func (c *hashRowContainer) matchJoinKey(buildRow, probeRow chunk.Row, probeHCtx *hashContext) (ok bool, err error) {
+
```
Comment (XuHuaiyu) [inline executor/hash_table.go]: This line is useless.

File: executor/hash_table.go [hunk 8]
```diff
@@ -14,31 +14,199 @@
...
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/sessionctx/stmtctx"
+	"github.com/pingcap/tidb/types"
 	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/codec"
+	"github.com/pingcap/tidb/util/memory"
+)
+
+const (
+	// estCountMaxFactor defines the factor of maxStatCount with maxChunkSize.
```
Comment (XuHuaiyu) [inline executor/hash_table.go]: 1. What does maxStatCount mean?
2. What does estCountMax mean?

File: executor/hash_table.go [hunk 9]
```diff
@@ -14,31 +14,199 @@
...
+	// estCountMax is maxChunkSize * estCountMaxFactor.
+	// Set this threshold to prevent innerEstCount being too large and causing a performance regression.
+	estCountMaxFactor = 10 * 1024
+
+	// estCountMinFactor defines the factor of statCountMin with maxChunkSize.
```
Comment (XuHuaiyu) [inline executor/hash_table.go]: ditto

File: executor/hash_table.go [hunk 10]
```diff
@@ -14,31 +14,200 @@
...
+)
+
+const (
+	// estCountMaxFactor defines the factor of estCountMax with maxChunkSize.
+	// estCountMax is maxChunkSize * estCountMaxFactor, the maximum threshold of estCount.
+	// if estCount is larger than estCountMax, set estCount to estCountMax.
+	// Set this threshold to prevent innerEstCount being too large and causing a performance and memory regression.
+	estCountMaxFactor = 10 * 1024
+
+	// estCountMinFactor defines the factor of statCountMin with maxChunkSize.
```
Comment (XuHuaiyu) [inline executor/hash_table.go]: s/ statCountMin/ estCountMin ?

**Pull request merged**